### ISSUE TITLE
Use of docker socket proxy no longer working

### ISSUE THREAD

**Issue by arcoast** [tale/headplane#30]
Just got around to updating headscale to beta2 and noticed that I can't use the docker socket proxy I was using previously, works fine if use wihout the proxy
```
2024-08-25T21:11:45.950464494Z 2024-08-25T21:11:45.949Z (INFO) [INTG] Loading integration: docker
2024-08-25T21:11:45.950724702Z 2024-08-25T21:11:45.950Z (INFO) [INTG] Using container: headscale
2024-08-25T21:11:45.950857410Z 2024-08-25T21:11:45.950Z (INFO) [INTG] Checking API: tcp://docker-socket-proxy:2375
2024-08-25T21:11:45.953229615Z 2024-08-25T21:11:45.953Z (ERRO) [INTG] Failed to connect to Docker API
2024-08-25T21:11:45.953254075Z 2024-08-25T21:11:45.953Z (ERRO) [INTG] Integration docker is not available
2024-08-25T21:11:45.953414375Z 2024-08-25T21:11:45.953Z (INFO) [CTXT] Starting Headplane with Context
2024-08-25T21:11:45.953446356Z 2024-08-25T21:11:45.953Z (INFO) [CTXT] HEADSCALE_URL: https://headscale.DOMAIN.COM
2024-08-25T21:11:45.953473469Z 2024-08-25T21:11:45.953Z (INFO) [CTXT] Integration: None
2024-08-25T21:11:45.953513737Z 2024-08-25T21:11:45.953Z (INFO) [CTXT] Config: Found 
2024-08-25T21:11:45.953610982Z 2024-08-25T21:11:45.953Z (INFO) [CTXT] OIDC: Configured
```

My compose file

```yaml
    headplane:
        image: ghcr.io/tale/headplane
        container_name: headplane
        networks:
            - traefik
        # ports:
        #     - 3000:3000
        environment:
            - COOKIE_SECRET=${HEADPLANE_COOKIE_SECRET}
            - HEADSCALE_URL=${HEADPLANE_HEADSCALE_URL}
            - HOST=${HEADPLANE_HOST}
            - PORT=${HEADPLANE_PORT}
            - CONFIG_FILE=${HEADPLANE_CONFIG_FILE}
            - ACL_FILE=${HEADPLANE_ACL_FILE}
            - HEADSCALE_CONFIG_UNSTRICT=${HEADPLANE_HEADSCALE_CONFIG_UNSTRICT}
            - DOCKER_SOCK=${HEADPLANE_DOCKER_SOCK}
            - HEADSCALE_CONTAINER=${HEADPLANE_HEADSCALE_CONTAINER}
            - OIDC_ISSUER=${HEADPLANE_OIDC_ISSUER}
            - OIDC_CLIENT_ID=${HEADPLANE_OIDC_CLIENT_ID}
            - OIDC_CLIENT_SECRET=${HEADPLANE_OIDC_CLIENT_SECRET}
            - ROOT_API_KEY=${HEADPLANE_ROOT_API_KEY}
            - DISABLE_API_KEY_LOGIN=${HEADPLANE_DISABLE_API_KEY_LOGIN}            
            - HEADSCALE_INTEGRATION=${HEADPLANE_HEADSCALE_INTEGRATION}
        volumes:
            - ${CONFIG}/headscale/headscale:/var/lib/headscale 
            - ${CONFIG}/headscale:/etc/headscale
        restart: unless-stopped
```

My `.env` file

```bash
HEADPLANE_COOKIE_SECRET=xxxxxxxxxxxxxxxxx
HEADPLANE_HEADSCALE_URL=https://headscale.${TRAEFIK_DOMAIN}
HEADPLANE_HOST=0.0.0.0
HEADPLANE_PORT=3000
HEADPLANE_CONFIG_FILE=/etc/headscale/config.yml
HEADPLANE_ACL_FILE=/etc/headscale/acl.jsonc
HEADPLANE_HEADSCALE_CONFIG_UNSTRICT=false

# Docker
HEADPLANE_HEADSCALE_INTEGRATION=docker
HEADPLANE_HEADSCALE_CONTAINER=headscale
HEADPLANE_DOCKER_SOCK="tcp://docker-socket-proxy:2375"

# OIDC
HEADPLANE_OIDC_ISSUER=https://login.${TRAEFIK_DOMAIN}
HEADPLANE_OIDC_CLIENT_ID=headscale
HEADPLANE_OIDC_CLIENT_SECRET=xxxxxxxxxxxxxxx
HEADPLANE_ROOT_API_KEY=xxxxxxxxxxxxxx
HEADPLANE_DISABLE_API_KEY_LOGIN=true
```

My docker socket proxy config for reference

```yaml
services:
    docker-socket-proxy:
        image: lscr.io/linuxserver/socket-proxy
        container_name: docker-socket-proxy
        networks:
            - traefik
        # ports:
        #     - 2375:2375
        environment:
            - TZ=${TZ}
            - CONTAINERS=1
            - IMAGES=1
        volumes:
            - /var/run/docker.sock:/var/run/docker.sock:ro
        restart: unless-stopped
```

Confirming I can access the docker socket proxy from headplane 

```
docker exec headscale apk add curl jq
```

and 

```json
docker exec headplane curl -s http://docker-socket-proxy:2375/v1.30/version | jq
{
  "Platform": {
    "Name": "Docker Engine - Community"
  },
  "Components": [
    {
      "Name": "Engine",
      "Version": "24.0.9",
      "Details": {
        "ApiVersion": "1.43",
        "Arch": "amd64",
        "BuildTime": "2024-02-01T00:49:16.000000000+00:00",
        "Experimental": "false",
        "GitCommit": "fca702d",
        "GoVersion": "go1.20.13",
        "KernelVersion": "6.1.99-Unraid",
        "MinAPIVersion": "1.12",
        "Os": "linux"
      }
    },
    {
      "Name": "containerd",
      "Version": "v1.7.13",
      "Details": {
        "GitCommit": "7c3aca7a610df76212171d200ca3811ff6096eb8"
      }
    },
    {
      "Name": "runc",
      "Version": "1.1.12",
      "Details": {
        "GitCommit": "v1.1.12-0-g51d5e94"
      }
    },
    {
      "Name": "docker-init",
      "Version": "0.19.0",
      "Details": {
        "GitCommit": "de40ad0"
      }
    }
  ],
  "Version": "24.0.9",
  "ApiVersion": "1.43",
  "MinAPIVersion": "1.12",
  "GitCommit": "fca702d",
  "GoVersion": "go1.20.13",
  "Os": "linux",
  "Arch": "amd64",
  "KernelVersion": "6.1.99-Unraid",
  "BuildTime": "2024-02-01T00:49:16.000000000+00:00"
}
```

**Comment by tale** [comment]
My question is are you sure this isn't a DNS issue of any sort. Node.js has notoriously had this issue before and I'm wondering if passing in the raw IP resolves this issue.

**Comment by arcoast** [comment]
I'll investigate a bit and check it's not DNS, but it's using docker native DNS rather than any upstream DNS, ie I'm referencing the container name `docker-socket-proxy` rather than any specific IP address and it's never a fixed IP.  I guess an easy way to test would be to expose the port on a bridge network to host and use the host IP.

Let me do some more testing and I'll post back later or over the weekend.  

FWIW the above was how it was configured before and was working fine.

**Comment by arcoast** [comment]
One other thought, the docker socket proxy has very granular permissions, starting pretty restrictive, which just made me think.

1.  What's the mechanism used to check the viability of the endpoint?  
    - Could it be failing the check?
2.  What actions does it perform via the API so I can check these are whitelisted?

In the meantime I'll open it up fully to test tomorrow (It's quite late here now)

**Comment by arcoast** [comment]
Made the socket proxy as open as open can be.

```yaml
services:
    docker-socket-proxy:
        image: lscr.io/linuxserver/socket-proxy
        container_name: docker-socket-proxy
        networks:
            - traefik
        ports:
            - 2375:2375
        environment:
            - TZ=${TZ}
            - ALLOW_START=1 #optional
            - ALLOW_STOP=1 #optional
            - ALLOW_RESTARTS=1 #optional
            - AUTH=1 #optional
            - BUILD=1 #optional
            - COMMIT=1 #optional
            - CONFIGS=1 #optional
            - CONTAINERS=1 #optional
            - DISABLE_IPV6=1 #optional
            - DISTRIBUTION=1 #optional
            - EVENTS=1 #optional
            - EXEC=1 #optional
            - IMAGES=1 #optional
            - INFO=1 #optional
            - NETWORKS=1 #optional
            - NODES=1 #optional
            - PING=1 #optional
            - POST=1 #optional
            - PLUGINS=1 #optional
            - SECRETS=1 #optional
            - SERVICES=1 #optional
            - SESSION=1 #optional
            - SWARM=1 #optional
            - SYSTEM=1 #optional
            - TASKS=1 #optional
            - VERSION=1 #optional
            - VOLUMES=1 #optional
        volumes:
            - /var/run/docker.sock:/var/run/docker.sock:ro
        restart: unless-stopped
```

Still no dice
```
2024-09-26T21:54:58.340902156Z 2024-09-26T21:54:58.340Z (INFO) [INTG] Loading integration: docker
2024-09-26T21:54:58.341112857Z 2024-09-26T21:54:58.341Z (INFO) [INTG] Using container: headscale
2024-09-26T21:54:58.341250742Z 2024-09-26T21:54:58.341Z (INFO) [INTG] Checking API: tcp://docker-socket-proxy:2375
2024-09-26T21:54:58.343371280Z 2024-09-26T21:54:58.343Z (ERRO) [INTG] Failed to connect to Docker API
2024-09-26T21:54:58.343414225Z 2024-09-26T21:54:58.343Z (ERRO) [INTG] Integration docker is not available
2024-09-26T21:54:58.343515608Z 2024-09-26T21:54:58.343Z (INFO) [CTXT] Starting Headplane with Context
2024-09-26T21:54:58.343542883Z 2024-09-26T21:54:58.343Z (INFO) [CTXT] HEADSCALE_URL: https://headscale.DOMAIN.COM
2024-09-26T21:54:58.343574413Z 2024-09-26T21:54:58.343Z (INFO) [CTXT] Integration: None
2024-09-26T21:54:58.343604956Z 2024-09-26T21:54:58.343Z (INFO) [CTXT] Config: Found 
2024-09-26T21:54:58.343712100Z 2024-09-26T21:54:58.343Z (INFO) [CTXT] OIDC: Configured
2024-09-26T21:54:58.349457987Z [remix-serve] http://localhost:3000 (http://0.0.0.0:3000)
```

**Comment by tale** [comment]
Let me enable a build with debug logs for you so that I can get the exact stack trace you encounter. Praying this helps me, since I can't reproduce your exact environment effectively.

**Comment by arcoast** [comment]
> Let me enable a build with debug logs for you so that I can get the exact stack trace you encounter. Praying this helps me, since I can't reproduce your exact environment effectively.

No worries, just ping me when you want me to test.

**Comment by tale** [comment]
I'll have a basic build out in a couple minutes with debug, but I think I identified the issue anyways. In the code I do `url.protocol = 'http:'` in order to make it work with fetch and it appears that it never worked the way I intended. Anyways, the edge image is currently building, can you try running it with `DEBUG=true` and see if the issue is solved or you can pull a better error?

**Comment by arcoast** [comment]
So here's the error I'm getting now with `DEBUG=true` on the `edge` image.

```
2024-10-03T18:59:07.187210270Z 2024-10-03T18:59:07.184Z (INFO) [CTXT] Debug mode is enabled! Logs will spam a lot.
2024-10-03T18:59:07.187454587Z 2024-10-03T18:59:07.187Z (INFO) [CTXT] Please disable debug mode in production.
2024-10-03T18:59:07.187577285Z 2024-10-03T18:59:07.187Z (DEBG) [CTXT] Checking config at /etc/headscale/config.yml
2024-10-03T18:59:07.187984738Z 2024-10-03T18:59:07.187Z (DEBG) [CFGX] Loading Headscale configuration from /etc/headscale/config.yml
2024-10-03T18:59:07.206741919Z 2024-10-03T18:59:07.206Z (DEBG) [CFGX] Attempting to parse Headscale configuration
2024-10-03T18:59:07.211717205Z 2024-10-03T18:59:07.211Z (DEBG) [CTXT] Checking write access to /etc/headscale/config.yml
2024-10-03T18:59:07.211992344Z 2024-10-03T18:59:07.211Z (INFO) [INTG] Loading integration: docker
2024-10-03T18:59:07.212152093Z 2024-10-03T18:59:07.212Z (DEBG) [INTG] Checking Docker integration availability
2024-10-03T18:59:07.212213252Z 2024-10-03T18:59:07.212Z (INFO) [INTG] Using container: headscale
2024-10-03T18:59:07.212359152Z 2024-10-03T18:59:07.212Z (INFO) [INTG] Checking API: http://docker-socket-proxy:2375
2024-10-03T18:59:07.241909960Z 2024-10-03T18:59:07.238Z (ERRO) [INTG] Failed to load integration docker: InvalidArgumentError: Invalid URL protocol: the URL must start with `http:` or `https:`.
2024-10-03T18:59:07.241939846Z     at parseURL (/app/node_modules/.pnpm/undici@6.19.2/node_modules/undici/lib/core/util.js:139:13)
2024-10-03T18:59:07.241942234Z     at Object.parseOrigin (/app/node_modules/.pnpm/undici@6.19.2/node_modules/undici/lib/core/util.js:206:9)
2024-10-03T18:59:07.241943813Z     at new Client (/app/node_modules/.pnpm/undici@6.19.2/node_modules/undici/lib/dispatcher/client.js:223:23)
2024-10-03T18:59:07.241945285Z     at Object.isAvailable (file:///app/build/server/index.js?t=1727944182000:165:25)
2024-10-03T18:59:07.241946773Z     at processTicksAndRejections (node:internal/process/task_queues:95:5)
2024-10-03T18:59:07.241948241Z     at loadIntegration (file:///app/build/server/index.js?t=1727944182000:450:17)
2024-10-03T18:59:07.241949846Z     at loadContext (file:///app/build/server/index.js?t=1727944182000:752:18)
2024-10-03T18:59:07.241951295Z     at file:///app/build/server/index.js?t=1727944182000:851:1 {
2024-10-03T18:59:07.241952767Z   code: 'UND_ERR_INVALID_ARG'
2024-10-03T18:59:07.241954271Z }
2024-10-03T18:59:07.242071961Z 2024-10-03T18:59:07.241Z (DEBG) [CTXT] Checking OIDC configuration
2024-10-03T18:59:07.242081118Z 2024-10-03T18:59:07.241Z (DEBG) [CTXT] API Key Login Enabled: false
2024-10-03T18:59:07.242083600Z 2024-10-03T18:59:07.242Z (DEBG) [CTXT] Checking ROOT_API_KEY and falling back to API_KEY
2024-10-03T18:59:07.242156265Z 2024-10-03T18:59:07.242Z (DEBG) [CTXT] Checking OIDC environment variables
2024-10-03T18:59:07.242190378Z 2024-10-03T18:59:07.242Z (DEBG) [CTXT] Issuer: https://authelia.DOMAIN.COM
2024-10-03T18:59:07.242302343Z 2024-10-03T18:59:07.242Z (DEBG) [CTXT] Client: headscale
2024-10-03T18:59:07.242310207Z 2024-10-03T18:59:07.242Z (INFO) [CTXT] Starting Headplane with Context
2024-10-03T18:59:07.242327770Z 2024-10-03T18:59:07.242Z (INFO) [CTXT] HEADSCALE_URL: https://headscale.DOMAIN.COM
2024-10-03T18:59:07.242406755Z 2024-10-03T18:59:07.242Z (INFO) [CTXT] Integration: None
2024-10-03T18:59:07.242433288Z 2024-10-03T18:59:07.242Z (INFO) [CTXT] Config: Found 
2024-10-03T18:59:07.242449051Z 2024-10-03T18:59:07.242Z (INFO) [CTXT] OIDC: Configured
2024-10-03T18:59:07.246561041Z [remix-serve] http://localhost:3000 (http://0.0.0.0:3000)
```

**Comment by tale** [comment]
Try this `edge` build once it's done.
https://github.com/tale/headplane/actions/runs/11168279805/job/31046321214

Remember you'll need to force pull it again.

**Comment by arcoast** [comment]
Yep, that's working now.

```
2024-10-03T19:25:36.838686782Z 2024-10-03T19:25:36.837Z (INFO) [CTXT] Debug mode is enabled! Logs will spam a lot.
2024-10-03T19:25:36.838876323Z 2024-10-03T19:25:36.838Z (INFO) [CTXT] Please disable debug mode in production.
2024-10-03T19:25:36.839030049Z 2024-10-03T19:25:36.838Z (DEBG) [CTXT] Checking config at /etc/headscale/config.yml
2024-10-03T19:25:36.839308558Z 2024-10-03T19:25:36.839Z (DEBG) [CFGX] Loading Headscale configuration from /etc/headscale/config.yml
2024-10-03T19:25:36.856511588Z 2024-10-03T19:25:36.856Z (DEBG) [CFGX] Attempting to parse Headscale configuration
2024-10-03T19:25:36.862024659Z 2024-10-03T19:25:36.861Z (DEBG) [CTXT] Checking write access to /etc/headscale/config.yml
2024-10-03T19:25:36.862258848Z 2024-10-03T19:25:36.862Z (INFO) [INTG] Loading integration: docker
2024-10-03T19:25:36.862403892Z 2024-10-03T19:25:36.862Z (DEBG) [INTG] Checking Docker integration availability
2024-10-03T19:25:36.862447107Z 2024-10-03T19:25:36.862Z (INFO) [INTG] Using container: headscale
2024-10-03T19:25:36.862573449Z 2024-10-03T19:25:36.862Z (INFO) [INTG] Checking API: http://docker-socket-proxy:2375
2024-10-03T19:25:36.889287022Z 2024-10-03T19:25:36.889Z (INFO) [INTG] Loaded integration: docker
2024-10-03T19:25:36.889579179Z 2024-10-03T19:25:36.889Z (DEBG) [CTXT] Checking OIDC configuration
2024-10-03T19:25:36.889607536Z 2024-10-03T19:25:36.889Z (DEBG) [CTXT] API Key Login Enabled: false
2024-10-03T19:25:36.889624077Z 2024-10-03T19:25:36.889Z (DEBG) [CTXT] Checking ROOT_API_KEY and falling back to API_KEY
2024-10-03T19:25:36.889784652Z 2024-10-03T19:25:36.889Z (DEBG) [CTXT] Checking OIDC environment variables
2024-10-03T19:25:36.889811517Z 2024-10-03T19:25:36.889Z (DEBG) [CTXT] Issuer: https://authelia.DOMAIN.COM
2024-10-03T19:25:36.889888302Z 2024-10-03T19:25:36.889Z (DEBG) [CTXT] Client: headscale
2024-10-03T19:25:36.889927019Z 2024-10-03T19:25:36.889Z (INFO) [CTXT] Starting Headplane with Context
2024-10-03T19:25:36.889946375Z 2024-10-03T19:25:36.889Z (INFO) [CTXT] HEADSCALE_URL: https://headscale.DOMAIN.COM
2024-10-03T19:25:36.890066880Z 2024-10-03T19:25:36.890Z (INFO) [CTXT] Integration: Docker
2024-10-03T19:25:36.890091122Z 2024-10-03T19:25:36.890Z (INFO) [CTXT] Config: Found 
2024-10-03T19:25:36.890104917Z 2024-10-03T19:25:36.890Z (INFO) [CTXT] OIDC: Configured
2024-10-03T19:25:36.894451122Z [remix-serve] http://localhost:3000 (http://0.0.0.0:3000)
```

**Comment by tale** [comment]
Ok perfect, I'm going to cut an official release with the debug logging as an option and this fix. Thank you for the help with debugging it.

**Comment by arcoast** [comment]
No worries  matey I should be thanking you for looking at it and fixing the issue!

So thanks!

Will leave you to close the issue with the release.

**Comment by tale** [comment]
Fixed in 0.3.0 🙂
